### PR TITLE
Sketch first part of the lab exercise

### PR DESCRIPTION
**Pull request by NicMcPhee**
This draft write-up gets through what I had envisioned as the "first phase", where we would make a separate contact info page for each person in the lab. I don't expect this will lead to any merge conflicts, but we will have rejected pushes and the need to merge pulls. In my head this was simple and short, but it's gotten kind of long, and I'm not sure if I'm writing too much or too little. Feedback on that would be greatly appreciated.

My idea for the second phase was that we would collectively generate a single contact info document. My assumption is that this will generate merge conflicts, and we'll thus get to deal with them. I'm not entirely sure we'll finish the _first_ phase in the first two hour lab, though, so I'm a little uncertain as to how to proceed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,3 +26,134 @@ a single-page app that goes through the major features of `git`
...
+Everyone in the class will have access to it, though, so you should
+put some thought in what you want to share.
+
+# Create a contact info page for (each of) you
+
+This repository has a `contact_info` directory, and in this part of the
+lab we'll create a new file in that directory for every person in the
+class, containing their personal contact info.
+
+:question: KK – do we want to have person X type up person Y's info
```
Comment (kklamberty) [inline README.md]: I think your idea is a good one. It will also cause people to have to talk and listen... both useful skills for communication.

**Comment by kklamberty** [comment]
It's a little weird reading markdown this way in order to give feedback on the lab. I imagine that for code it makes a bit more sense, but I wonder if the lack of color coding here will mean that people are looking at code in webstorm rather than here? Or, was it your intent that I do this through webstorm... does it support that?

File: README.md [hunk 2]
```diff
@@ -147,13 +147,97 @@ this is easy:
...
 repository. This is the potentially tricky part. :smile:
 
-If no one else was working on this repository, it would be easy:
+If no one else is working on this branch in this repository (which is
+likely in this case since you're working on a special branch you created
+for this contact info document) all you'll need to do is _push_ your
+work up to the Github repository:
+* `git push` on the command line
+* `VCS -> Git -> Push…` in WebStorm
+This will push all commits on your current branch so a branch with the
```
Comment (kklamberty) [inline README.md]: "to" a branch rather than "so" a branch

File: README.md [hunk 3]
```diff
@@ -1,2 +1,243 @@
...
+(and [a later post arguing that GitFlow is a Very Bad Idea](http://endoflineblog.com/gitflow-considered-harmful))
+* [Git Essential Training](https://www.lynda.com/Git-tutorials/Git-Essential-Training/100222-2.html),
+an on-line Lynda.com course; it's free if you first authenticate
+with your U of M credentials via http://lynda.umn.edu
+
+The discussion below assumes that people are paired up in the lab, and
+there will be times where one person will need to log out and let their
+partner login so that everyone has a chance to go through all the
+activities.
+
```
Comment (kklamberty) [inline README.md]: Joe used one of those special incognito windows (or similar) when we worked on something for GitHub Classroom, and that also worked just fine. That might be faster.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,243 @@
...
+`VCS -> Git -> Branches…` dialogue, choosing the `New Branch` option.
+
+## Creating a page
+
+Once you've created the branch it's time to create your partner's
+contact info page.
+
+Each contact info page should:
+
+* Be named `<your_name>.md`, e.g., `Nic_McPhee.md`
```
Comment (kklamberty) [inline README.md]: It won't be "your" name, but it will be the partner's name. It's the name of the person whose contact info will be there.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,243 @@
...
+
+After a page is created, you need to share that with the class by
+adding it to the repository on Github.
+
+Because you've created your own branch to work in, the beginning of
+this is easy:
+
+* Stage the new file so `git` knows you want to commit it
+   * `git add <filename>` on the command line
+   * `VCS -> Git -> Add` in WebStorm (although if you used WebStorm to
```
Comment (kklamberty) [inline README.md]: Maybe the part about add in webstorm should be moved or also included above so that people know the answer to that question if it comes up.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,243 @@
...
+   create the file, it probably asked you if you wanted to add it to
+   the set of files that were being tracked by `git`, so you may not
+   need to explicitly add it)
+* Commit the change _to your local copy of the repository_
+   * `git commit` on the command line.
+      * This will bring up a (command line) editor for you to enter
+      your commit message. By default this is `vi/vim` for most people;
+      ask for help if you find yourself trapped in `vim` and can't seem
+      to escape.
+      * You can also use `git commit -m "Your cool commit message"` to
```
Comment (kklamberty) [inline README.md]: Does -m mean you can only write a little? Or, does that just encourage people to only write a little? If we are going to strongly discourage this, I think we should have a page that explains how to change your default editor. I think pico/nano is far simpler and would still allow for some more formatting and such in a comment.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,243 @@
...
+      avoid being sent to an editor. We _strongly_ discourage this,
+      however, as people rarely enter useful one-line commit messages.
+    * `VCS -> Commit Changes…` in WebStorm. This gives you a nice
+    dialogue where you can select which files you actually
```
Comment (kklamberty) [inline README.md]: "This is useful because you may have made a few changes that are related to different issues - you can make sure your commit makes sense by commenting separately for unrelated changes." Or, something like that. I think this matters enough to write it in the text of the lab.

File: README.md (hunk 6 above)
Comment (kklamberty) [inline README.md]: I should clarify that I know that -m is for adding your message - I just don't know if there are restrictions on how long the message can be.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,243 @@
...
+on a particular feature from other on-going work on the project.
+Creating and working in your own feature branch allows you to commit
+and share incomplete work without breaking the project for other people.
+
+To illustrate the use of branches (which arguably aren't really needed
+for this simple project), we'll create a new branch for the creation
+of each contact info page. So in the example above, the first thing
+that Pat would do would be to create a branch for the creation of
+Chris's contact info page.
+
```
Comment (kklamberty) [inline README.md]: This is my test of reviewing

File: README.md [hunk 9]
```diff
@@ -1,2 +1,243 @@
...
+In WebStorm, you can accomplish this through the
+`VCS -> Git -> Branches…` dialogue, choosing the `New Branch` option.
+
+## Creating a page
+
+Once you've created the branch it's time to create your partner's
+contact info page.
+
+Each contact info page should:
```
Comment (kklamberty) [inline README.md]: ssdfasdf

File: README.md (hunk 1 above)
Comment (NicMcPhee) [inline README.md]: :)

**Comment by NicMcPhee** [comment]
> It's a little weird reading markdown this way in order to give feedback on the lab. I imagine that for code it makes a bit more sense, but I wonder if the lack of color coding here will mean that people are looking at code in webstorm rather than here? Or, was it your intent that I do this through webstorm... does it support that?

I don't think that WebStorm provides any support for this kind of review, so it will have to happen on GitHub.

Actually the code coloring *does* show up when doing a review in Github. I think it's not happening here because this is all Markdown, but if you look at [this review](https://github.com/NicMcPhee/a-star-search/pull/18) you can see that the Clojure code is still being colored.

That said, when I did this for class, I did often find myself needing to checkout the code and play with it. There are times (for me) when reading coding isn't enough to convince me that it does (or doesn't) work, and I had to actually run things to see. Well written tests ought to really help with that, though, so maybe this will provide some incentive on that front?

File: README.md (hunk 3 above)
Comment (NicMcPhee) [inline README.md]: That would work on for the web, but it won't work for editing in WebStorm locally, committing, pushing, etc.

There are ways to hack around that, but they're ugly and error prone, so having people log out/in may be the safest/simplest option.

File: README.md (hunk 4 above)
Comment (NicMcPhee) [inline README.md]: Good catch – thanks.

File: README.md (hunk 6 above)
Comment (NicMcPhee) [inline README.md]: I don't think there are restrictions on how long a message can be with `-m`, but as a practical matter people don't write longer (multi-line, or even multi-paragraph) commit messages with `-m`.

We should show people how to set their default editor to pico/nano. It may not matter a lot, though, because most people will probably use WebStorm most of the time, right?

File: README.md (hunk 7 above)
Comment (NicMcPhee) [inline README.md]: Nice.

**Comment by NicMcPhee** [comment]
@joethe – you have any thoughts on this?

File: README.md (hunk 5 above)
Comment (NicMcPhee) [inline README.md]: Good catch. I essentially copied that text up to where they create the file, so they're warned that might happen.

**Comment by NicMcPhee** [comment]
@kklamberty – I think this addresses all of your concerns. Note how several of your comments are now "gone" because I changed the code that was referred to in the comment. If you click the little "Show outdated" link it will reveal that bit of code and your comment. I think there are some nice tools for showing what's changed since your last review, so you can more easily focus on just the new changes.

File: README.md (hunk 3 above)
Comment (joethe) [inline README.md]: I agree. There are ways to commit "as" someone else, but that's a bit of a hassle if you're not used to doing it.

**that said:** This lab doesn't seem to make any mention of actually setting up their local git configurations, which they will likely be prompted to do if they haven't used git before. This is, essentially, just setting your name and email, but it's an important step.

Comment (NicMcPhee) [inline README.md]: Good catch - I tend to forget things like that because I've done them so long ago.
